scripts/src/utils/nextBlockHash.ts

```diff
@@ -35,111 +39,206 @@ export async function nextBlockHash(
 const cachedBlocks: Map<number, providers.Block> =
   new Map<number, providers.Block>()
 
+
+let latestBlockCache: providers.Block | undefined = undefined
+async function getLatestBlock(
+  provider: providers.Provider
+): Promise<providers.Block> {
+  if (latestBlockCache === undefined) {
+    latestBlockCache = await provider.getBlock("latest")
+  }
+  return latestBlockCache
+}
+
 async function getBlock(
   blockNumber: number,
   provider: providers.BaseProvider
-): Promise<providers.Block> {
+): Promise<providers.Block | undefined> {
+  if (blockNumber < 0) {
+    return undefined
+  }
   if (!cachedBlocks.has(blockNumber)) {
-    cachedBlocks.set(blockNumber, await provider.getBlock(blockNumber))
+    const maybeBlock = await provider.getBlock(blockNumber)
+    if (maybeBlock === null) {
+      // future block?
+      return undefined
+    }
+    cachedBlocks.set(blockNumber, maybeBlock)
   }
   const block = cachedBlocks.get(blockNumber)
   if (block === undefined) {
-    console.log("WARNING: expected cached block but didn't get one")
+    console.warn("Expected cached block but didn't get one")
     return await provider.getBlock(blockNumber)
   }
   return block
 }
 
-/**
- * - if no block, set block to latest block
- * - find time difference between block.timestamp and targetTime
- * - if time difference is 0, return block's hash
- * - if previousBlock
- *   - if block and previous block "surround" targetTime && are 1 block apart, return the later block hash of block and previous block
- * - set newBlock using time difference
- * - if newBlock == block,
- *   - if block is earlier than target time, set newBlock to 1 block later
- *   - else, set newBlock to 1 block earlier
- * - recurse with block: newBlock, previousBlock: block
- */
 export async function closestEarlierBlockHash(
```

Comment:
I'd love to see a pseudocode version of this algorithm -- a lot going on here it seems